geode-redis/src/main/java/org/apache/geode/redis/internal/data/RedisHash.java

```diff
@@ -203,49 +279,122 @@ public int hstrlen(ByteArrayWrapper field) {
   }
 
-  public Pair<BigInteger, List<Object>> hscan(Pattern matchPattern, int count, BigInteger cursor) {
-    List<Object> returnList = new ArrayList<Object>();
-    int size = hash.size();
-    BigInteger beforeCursor = new BigInteger("0");
-    int numElements = 0;
-    int i = -1;
-    for (Map.Entry<ByteArrayWrapper, ByteArrayWrapper> entry : hash.entrySet()) {
-      ByteArrayWrapper key = entry.getKey();
-      ByteArrayWrapper value = entry.getValue();
-      i++;
-      if (beforeCursor.compareTo(cursor) < 0) {
-        beforeCursor = beforeCursor.add(new BigInteger("1"));
+  public ImmutablePair<BigInteger, List<Object>> hscan(UUID clientID, Pattern matchPattern,
+      int count,
+      BigInteger cursorParameter) {
+
+    int startCursor = cursorParameter.intValue();
+
+    List<ByteArrayWrapper> keysToScan = getSnapShotOfKeySet(clientID);
+
+    Pair<Integer, List<Object>> resultsPair =
+        getResultsPair(keysToScan, startCursor, count, matchPattern);
+
+    List<Object> resultList = resultsPair.getRight();
+
+    Integer numberOfIterationsCompleted = resultsPair.getLeft();
+
+    int returnCursorValueAsInt =
+        getCursorValueToReturn(startCursor, numberOfIterationsCompleted, keysToScan);
+
+    if (returnCursorValueAsInt == 0) {
+      removeHSCANSnapshot(clientID);
+    }
+
+    return new ImmutablePair<>(BigInteger.valueOf(returnCursorValueAsInt),
+        resultList);
+  }
+
+  private void removeHSCANSnapshot(UUID clientID) {
+    this.hScanSnapShots.remove(clientID);
+    this.hScanSnapShotCreationTimes.remove(clientID);
+
+    if (this.hScanSnapShots.isEmpty()) {
+      shutDownHscanSnapshotScheduledRemoval();
+    }
+  }
+
+  @SuppressWarnings("unchecked")
+  private Pair<Integer, List<Object>> getResultsPair(List<ByteArrayWrapper> keysSnapShot,
+      int startCursor,
+      int count,
+      Pattern matchPattern) {
+
+    int indexOfKeys = startCursor;
+
+    List<ByteArrayWrapper> resultList = new ArrayList<>();
+
+    for (int index = startCursor; index < keysSnapShot.size(); index++) {
+
+      if ((index - startCursor) == count) {
+        break;
+      }
+
+      ByteArrayWrapper key = keysSnapShot.get(index);
+      indexOfKeys++;
+
+      ByteArrayWrapper value = hash.get(key);
+      if (value == null) {
         continue;
       }
 
       if (matchPattern != null) {
         if (matchPattern.matcher(key.toString()).matches()) {
-          returnList.add(key);
-          returnList.add(value);
-          numElements++;
+          resultList.add(key);
+          resultList.add(value);
         }
       } else {
-        returnList.add(key);
-        returnList.add(value);
-        numElements++;
+        resultList.add(key);
+        resultList.add(value);
       }
+    }
 
-      if (numElements == count) {
-        break;
-      }
+    Integer numberOfIterationsCompleted = indexOfKeys - startCursor;
+
+    Pair resultsPair = new ImmutablePair(numberOfIterationsCompleted, resultList);
```

Comment:
you could return this directly instead of having this extra assignment

Reply:
ok